### ISSUE TITLE
Question: how health checking works ? why my services are always unhealthy?

### ISSUE THREAD

**Issue by xinyu7git** [googleforgames/agones#1263]
```
[root@172-18-212-12 umonitor3]# kubectl get gs
NAME                      STATE       ADDRESS          PORT   NODE             AGE
transfer-producer-hzhk6   Unhealthy   172.18.178.255   8501   172-18-178-255   6m9s
```

**Comment by xinyu7git** [comment]
kind/question

**Comment by xinyu7git** [comment]
/kind/question

**Comment by markmandel** [comment]
Please let us know if there is anything missing from our documentation: https://agones.dev/site/docs/guides/health-checking/

**Comment by xinyu7git** [comment]
> Please let us know if there is anything missing from our documentation: https://agones.dev/site/docs/guides/health-checking/

I can't find how health checking works in docs. 

when i use `kubectl describe pod xxx`:
```
[root@172-18-212-12 umonitor3]# kubectl describe pod transfer-producer-6t52d
Name:         transfer-producer-6t52d
Namespace:    default
Priority:     0
Node:         172-18-178-251/172.18.178.251
Start Time:   Wed, 08 Jan 2020 17:13:19 +0800
Labels:       agones.dev/gameserver=transfer-producer-6t52d
              agones.dev/role=gameserver
Annotations:  agones.dev/container: transfer-producer
              agones.dev/sdk-version: 1.2.0
              cluster-autoscaler.kubernetes.io/safe-to-evict: false
              cni.projectcalico.org/podIP: 10.97.111.181/32
Status:       Running
IP:           10.97.111.181
IPs:
  IP:           10.97.111.181
Controlled By:  GameServer/transfer-producer-6t52d
Containers:
  transfer-producer:
    Container ID:  docker://169ae2664e201b3b45f53380ae9fd2b23844490571ae95d94641d86cb1e7e576
    Image:         hub.ucloudadmin.com/umonitor3/transfer-producer:202001081702
    Image ID:      docker-pullable://hub.ucloudadmin.com/umonitor3/transfer-producer@sha256:fbc16d0257f1bc331ba528516f0ec378dbe2847c5d411b27a734ea8cf75bbca3
    Ports:         8501/TCP, 8601/TCP
    Host Ports:    7160/TCP, 7528/TCP
    Command:
      /root/run.sh
    State:          Running
      Started:      Wed, 08 Jan 2020 17:18:42 +0800
    Last State:     Terminated
      Reason:       OOMKilled
      Exit Code:    137
      Started:      Wed, 08 Jan 2020 17:18:00 +0800
      Finished:     Wed, 08 Jan 2020 17:18:40 +0800
    Ready:          True
    Restart Count:  5
    Limits:
      cpu:     20m
      memory:  32Mi
    Requests:
      cpu:     20m
      memory:  32Mi
    Liveness:  http-get http://:8080/gshealthz delay=5s timeout=1s period=5s #success=1 #failure=3
    Environment:
      CLUSTER_NAME:          transfer-producer
      CONFIG_SERVICE_ADDR:   172.29.142.89:8080
      AGONES_SDK_GRPC_PORT:  9357
      AGONES_SDK_HTTP_PORT:  9358
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from empty (ro)
  agones-gameserver-sidecar:
    Container ID:  docker://3bfcb427344b78a29d94d1ad55f392322aa27da1040c31517172722c94e59532
    Image:         hub.ucloudadmin.com/google_containers/agones-sdk:1.2.0
    Image ID:      docker-pullable://hub.ucloudadmin.com/google_containers/agones-sdk@sha256:66e62202bbe9e0a1b7d0da35fbe9fbd7878f495b716dfd9faf5d97c8a61edb0a
    Port:          <none>
    Host Port:     <none>
    Args:
      --grpc-port=9357
      --http-port=9358
    State:          Running
      Started:      Wed, 08 Jan 2020 17:13:21 +0800
    Ready:          True
    Restart Count:  0
    Requests:
      cpu:     30m
    Liveness:  http-get http://:8080/healthz delay=3s timeout=1s period=3s #success=1 #failure=3
    Environment:
      GAMESERVER_NAME:  transfer-producer-6t52d
      POD_NAMESPACE:    default (v1:metadata.namespace)
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from agones-sdk-token-k6zk2 (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  empty:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  agones-sdk-token-k6zk2:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  agones-sdk-token-k6zk2
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason     Age                   From                     Message
  ----     ------     ----                  ----                     -------
  Normal   Scheduled  5m33s                 default-scheduler        Successfully assigned default/transfer-producer-6t52d to 172-18-178-251
  Normal   Pulled     5m31s                 kubelet, 172-18-178-251  Container image "hub.ucloudadmin.com/google_containers/agones-sdk:1.2.0" already present on machine
  Normal   Created    5m31s                 kubelet, 172-18-178-251  Created container agones-gameserver-sidecar
  Normal   Started    5m31s                 kubelet, 172-18-178-251  Started container agones-gameserver-sidecar
  Normal   Created    2m9s (x4 over 5m33s)  kubelet, 172-18-178-251  Created container transfer-producer
  Normal   Started    2m8s (x4 over 5m31s)  kubelet, 172-18-178-251  Started container transfer-producer
  Warning  Unhealthy  116s (x4 over 4m6s)   kubelet, 172-18-178-251  Liveness probe failed: HTTP probe failed with statuscode: 500
  Warning  BackOff    95s (x6 over 3m39s)   kubelet, 172-18-178-251  Back-off restarting failed container
  Normal   Pulled     12s (x6 over 5m33s)   kubelet, 172-18-178-251  Container image "hub.ucloudadmin.com/umonitor3/transfer-producer:202001081702" already present on machine
```

the message ` Liveness probe failed: HTTP probe failed with statuscode: 500` makes gameserver to be unhealthy state.  
In pod , there is another http server which binding to 8080 : 
```
[root@transfer-producer-6t52d ~]# ss -anplt
State       Recv-Q Send-Q                         Local Address:Port                                        Peer Address:Port
LISTEN      0      128                                127.0.0.1:9357                                                   *:*
LISTEN      0      128                                127.0.0.1:9358                                                   *:*
LISTEN      0      128                                       :::8080                                                  :::*
LISTEN      0      128                                       :::8501                                                  :::*                   users:(("transfer-produc",pid=11,fd=6))
LISTEN      0      128                                       :::8601                                                  :::*                   users:(("transfer-produc",pid=11,fd=7))
```
Liveness is using `http-get http://:8080/gshealthz ` and `http-get http://:8080/healthz`. this codes are not in my application, so i don't know why this happened.
```
[root@172-18-212-12 umonitor3]# kubectl describe pod transfer-producer-6t52d |grep Liveness
    Liveness:  http-get http://:8080/gshealthz delay=5s timeout=1s period=5s #success=1 #failure=3
    Liveness:  http-get http://:8080/healthz delay=3s timeout=1s period=3s #success=1 #failure=3
  Warning  Unhealthy  6m54s (x4 over 9m4s)  kubelet, 172-18-178-251  Liveness probe failed: HTTP probe failed with statuscode: 500
```
I don't know if I can make it clear， could you give me some suggests？

**Comment by aLekSer** [comment]
Just as a side notice. This is what looks strange to me :
```
    Last State:     Terminated
      Reason:       OOMKilled
```
Which is out of memory killed reason. You can first test your server locally to check that you do not over-consume memory:
https://agones.dev/site/docs/guides/client-sdks/local/

**Comment by markmandel** [comment]
> I can't find how health checking works in docs.

Sorry, I'm not sure what you mean here, I provided the link to the health checking documentation.

Could you rephrase this, maybe we're misunderstanding each other.

Something else to try is to `kubectl describe gs <name>` to see the events for that GameServer.
https://agones.dev/site/docs/guides/troubleshooting/

But looks like OOM might be the culprit. That container has restarted 5 times, which makes me think the crash is causing the GameServer to move to Unhealthy.

**Comment by xinyu7git** [comment]
ooooh, that's my fault, my pod was killed beause of oom. 
thanks. @markmandel  @aLekSer 

the 8080 port was exposed by sdk-server, i found codes in repo.
sdk-client used sdk.Health() func to connect to sdk-server and send `ping` message to sdk-server.

sdk-server used the `ping` message to judge my application's healthy. Is it right?

**Comment by markmandel** [comment]
> sdk-server used the ping message to judge my application's healthy. Is it right?

Yes, that is correct.

**Comment by Hades32** [comment]
Docs *really* should mention that your app is not allowed to use port 8080 to listen...
